README.md
Kipper
======

A server that updates GitHub pull request commit statuses from Jenkins builds.


## Prerequisites

A Jenkins server configured with a “branches” project. This project will run a
build for all new commits pushed to a branch. The project’s name should match
the GitHub repository name appended by “-branches”. For example:

	kipper-branches

Each build name will have the following format:

	branch-name-5ac92

The branch name, followed by a hyphen “-”, followed by the first five characters
of the commit SHA.


## Setup

Kipper runs a web server that listens for a GitHub webhook. Configure your
GitHub project with the following webhook settings:

	Payload URL: http://example.com/github/pull_request_event
	Content type: application/json
	Events: Pull request
	Active: true

Since GitHub will be sending webhook requests to Kipper, it must be publicly
accessible.

To run Kipper, several configuration parameters must be passed in via command
line arguments in order for it to communicate with Jenkins and update commit
statuses on GitHub:

	./kipper \
		--jenkins-url 'http://jenkins.example.com' \
		--jenkins-user-id 'username' \
		--jenkins-token 'jenkins-token' \
		--github-token 'github-token'

By default, Kipper will run on port 8000.


## Install

A binary built for Mac OS X is available on the [releases][1] page. Download the
binary and put it in your `PATH`.

To compile from source:

	$ cargo install --git https://github.com/teddywing/kipper.git --root /usr/local


## Uninstall

	$ cargo uninstall --root /usr/local kipper


## License
Copyright © 2017 Teddy Wing. Licensed under the GNU GPLv3+ (see the included
COPYING file).


[1]: https://github.com/teddywing/kipper/releases


CHANGELOG
CHANGELOG
=========

v0.0.3 (2017-11-22):
	Changes:

	* Listen on `0.0.0.0` instead of `127.0.0.1` to facilitate running the
	  server inside a Docker container.

v0.0.2 (2017-11-17):
	Fixes:

	* When a Jenkins build finished and Kipper finished polling, the final
	  GitHub status would be set as "pending" instead of "success" or
	  "failure". It now sends the correct status.
	* It was possible for Kipper to request Jenkins before a job
	  corresponding to the latest code in the pull request had been created.
	  An initial delay of 30 seconds is added before Kipper contacts Jenkins
	  to give it time to warm up.
	* Send the Jenkins job console URL instead of the job index URL with the
	  GitHub commit status request. Now, clicking on the status icon in
	  GitHub opens the console page so the build log is visible immediately.
	* Miscellaneous code refinements.

v0.0.1 (2017-11-13):
	* Initial release.


src/af83.rs
// Copyright © 2017 Teddy Wing
//
// This file is part of Kipper.
//
// Kipper is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Kipper is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Kipper. If not, see <http://www.gnu.org/licenses/>.

use pull_request::CommitRef;

pub fn job_name(commit_ref: &CommitRef) -> String {
    let (sha, _) = commit_ref.sha.split_at(5);

    format!("{}-{}", commit_ref.branch, sha)
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn job_name_is_branch_appended_by_commit_sha_prefix() {
        let commit_ref = CommitRef {
            owner: "sybil".to_owned(),
            repo: "sybil-system".to_owned(),
            sha: "159f8769b897ed7774700d0b2777def8ac838b8f".to_owned(),
            branch: "5912-make-logo-bigger".to_owned(),
        };

        assert_eq!(
            job_name(&commit_ref),
            "5912-make-logo-bigger-159f8"
        );
    }
}


src/github.rs
// Copyright © 2017 Teddy Wing
//
// This file is part of Kipper.
//
// Kipper is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Kipper is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Kipper. If not, see <http://www.gnu.org/licenses/>.

extern crate mockito;
extern crate reqwest;

use std::collections::HashMap;
use std::error::Error;
use std::fmt;

use self::reqwest::header::{Accept, Authorization, Bearer, qitem};

use pull_request::CommitRef;

#[cfg(not(test))]
const API_URL: &'static str = "https://api.github.com";

#[cfg(test)]
const API_URL: &'static str = mockito::SERVER_URL;

pub enum CommitStatus {
    Error,
    Failure,
    Pending,
    Success,
}

impl fmt::Display for CommitStatus {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            CommitStatus::Error => write!(f, "error"),
            CommitStatus::Failure => write!(f, "failure"),
            CommitStatus::Pending => write!(f, "pending"),
            CommitStatus::Success => write!(f, "success"),
        }
    }
}

pub fn update_commit_status(
    github_token: &String,
    commit_ref: &CommitRef,
    state: &CommitStatus,
    target_url: String,
    description: Option<String>,
    context: String,
) -> Result<(), Box<Error>> {
    let client = reqwest::Client::new();

    let mut params = HashMap::new();
    params.insert("state", state.to_string());
    params.insert("target_url", target_url);
    params.insert("context", context);

    if let Some(d) = description {
        params.insert("description", d);
    }

    let mut response = client.post(
            &format!(
                "{}/repos/{}/{}/statuses/{}",
                API_URL,
                commit_ref.owner,
                commit_ref.repo,
                commit_ref.sha
            )
        )
        .header(
            Accept(
                vec![qitem("application/vnd.github.v3+json".parse()?)]
            )
        )
        .header(
            Authorization(
                Bearer {
                    token: github_token.to_owned()
                }
            )
        )
        .json(&params)
        .send()?;

    debug!("{}", response.url());
    debug!("{}", response.status());
    debug!("{}", response.headers());
    debug!("{}", response.text()?);

    Ok(())
}


#[cfg(test)]
mod tests {
    use self::mockito::mock;

    use super::*;

    #[test]
    fn update_commit_status_makes_a_request_to_github() {
        let mock = mock("POST", "/repos/octocat/Hello-World/statuses/6dcb09b5b57875f334f61aebed695e2e4193db5e")
            .with_status(201)
            .create();

        let commit_ref = CommitRef {
            owner: "octocat".to_owned(),
            repo: "Hello-World".to_owned(),
            sha: "6dcb09b5b57875f334f61aebed695e2e4193db5e".to_owned(),
            branch: "not-used".to_owned(),
        };

        update_commit_status(
            &"token".to_owned(),
            &commit_ref,
            &CommitStatus::Success,
            "https://jenkins.example.com/job/octocat/3".to_owned(),
            None,
            "continuous-integration/jenkins".to_owned()
        ).expect("Failed to update commit status");

        mock.assert();
    }
}


src/jenkins.rs
// Copyright © 2017 Teddy Wing
//
// This file is part of Kipper.
//
// Kipper is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Kipper is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Kipper. If not, see <http://www.gnu.org/licenses/>.

// maybe wait a few seconds to be sure a Jenkins job was created. This happens at the caller.

// make request to [branch]-branches
// if it comes back successfully with a `builds` hash
// request all URLs in `builds`
//   if its `displayName` matches [branch]-commitsha{5}
//     check `result` ('SUCCESS', 'FAILURE', nonexistent)
//     update GitHub commit status
//       if pending
//           start a thread that checks every 30 seconds for the `result` and update GitHub commit status
//                 if time spent > 20 minutes
//                    set GH commit status to error (timeout)
//                 if `result` is successful or failed, update status and stop
//   set GH status to error (no job found)

// fn update_github_status(commit_ref)
// fn get_jobs(repo_name)
// fn af83 job name from commit_ref (separate af83 module)
// fn update_github_commit_status(status, message) (lives in GitHub module)
// fn request_job(url)
// fn result_from_job(payload)

extern crate json;
extern crate mockito;
extern crate reqwest;
extern crate url;

use std::error::Error;
use std::thread::sleep;
use std::time::{Duration, Instant};

use self::reqwest::header;
use self::url::Url;

use af83;
use github;
use pull_request::CommitRef;

#[derive(Debug, PartialEq, Eq)]
pub enum JobStatus {
    Success,
    Failure,
    Pending,
    Unknown,
}

impl JobStatus {
    fn commit_status(&self) -> github::CommitStatus {
        match *self {
            JobStatus::Success => github::CommitStatus::Success,
            JobStatus::Failure => github::CommitStatus::Failure,
            JobStatus::Pending => github::CommitStatus::Pending,
            JobStatus::Unknown => github::CommitStatus::Error,
        }
    }
}

pub struct Job {
    display_name: String,
    result: JobStatus,
}

impl Job {
    fn new(payload: String) -> Result<Job, Box<Error>> {
        let mut job = json::parse(payload.as_ref())?;

        Ok(
            Job {
                display_name: job["displayName"].take_string().unwrap_or_default(),
                result: result_from_job(job["result"].take_string()),
            }
        )
    }
}

pub fn find_and_track_build_and_update_status(
    commit_ref: CommitRef,
    jenkins_url: String,
    jenkins_user_id: &String,
    jenkins_token: &String,
    github_token: String,
) -> Result<(), Box<Error>> {
    let jenkins_client = jenkins_request_client(
        &jenkins_user_id,
        &jenkins_token
    )?;

    let jobs = get_jobs(
        &jenkins_url,
        &jenkins_client,
        commit_ref.repo.as_ref()
    )?;
    let t20_minutes = 60 * 20;

    for job_url in jobs {
        debug!("Looking for job: {}", job_url);

        let mut job = request_job(
            &jenkins_url,
            &jenkins_client,
            job_url.as_ref()
        )?;

        // Does `displayName` match
        if job_for_commit(&job, &commit_ref) {
            debug!("Job found: {}", job_url);

            // Start timer
            let now = Instant::now();

            let commit_status = job.result.commit_status();

            let job_console_url = jenkins_console_url_path(&job_url);

            github::update_commit_status(
                &github_token,
                &commit_ref,
                &commit_status,
                job_console_url.clone(),
                None,
                "continuous-integration/jenkins".to_owned()
            ).expect(
                format!(
                    "GitHub pending status update failed for {}/{} {}.",
                    commit_ref.owner,
                    commit_ref.repo,
                    commit_ref.sha
                ).as_ref()
            );

            while job.result == JobStatus::Pending {
                // loop
                // if timer > 20 minutes
                //   call github::update_commit_status with timeout error
                //   return
                // wait 30 seconds
                // call request_job again
                // if the status is different
                //   call github::update_commit_status
                //   stop

                debug!("Waiting for job to finish");

                if now.elapsed().as_secs() == t20_minutes {
                    github::update_commit_status(
                        &github_token,
                        &commit_ref,
                        &github::CommitStatus::Error,
                        job_console_url.clone(),
                        Some("The status checker timed out.".to_owned()),
                        "continuous-integration/jenkins".to_owned()
                    ).expect(
                        format!(
                            "GitHub timeout error status update failed for {}/{} {}.",
                            commit_ref.owner,
                            commit_ref.repo,
                            commit_ref.sha
                        ).as_ref()
                    );

                    return Ok(())
                }

                sleep(Duration::from_secs(30));

                let updated_job = request_job(
                    &jenkins_url,
                    &jenkins_client,
                    job_url.as_ref()
                ).expect(
                    format!("Failed to request job '{}'.", job_url).as_ref()
                );

                if job.result != updated_job.result {
                    github::update_commit_status(
                        &github_token,
                        &commit_ref,
                        &updated_job.result.commit_status(),
                        job_console_url.clone(),
                        None,
                        "continuous-integration/jenkins".to_owned()
                    ).expect(
                        format!(
                            "GitHub status update failed for {}/{} {}.",
                            commit_ref.owner,
                            commit_ref.repo,
                            commit_ref.sha
                        ).as_ref()
                    );

                    return Ok(())
                }

                job = updated_job;
            }

            return Ok(())
        }
    }

    Ok(())
}

pub fn auth_credentials(user_id: String, token: String) -> header::Basic {
    header::Basic {
        username: user_id,
        password: Some(token),
    }
}

pub fn get_jobs(
    jenkins_url: &String,
    client: &reqwest::Client,
    repo_name: &str
) -> Result<Vec<String>, Box<Error>> {
    let mut response = client.get(
        &format!("{}/job/{}-branches/api/json", jenkins_url, repo_name)
    ).send()?;

    let body = response.text()?;

    let jobs = json::parse(body.as_ref())?;

    Ok(
        jobs["builds"].members()
            .map(|job| {
                job["url"].clone().take_string().unwrap_or_default()
            })
            .collect::<Vec<String>>()
    )
}

pub fn request_job(
    jenkins_url: &String,
    client: &reqwest::Client,
    url: &str
) -> Result<Job, Box<Error>> {
    let url = Url::parse(url.as_ref())?;

    let mut response = client.get(
        &format!("{}{}/api/json", jenkins_url, url.path())
    ).send()?;

    let body = response.text()?;

    let job = Job::new(body)?;

    Ok(job)
}

// Does the `commit_ref` correspond to the job?
pub fn job_for_commit(job: &Job, commit_ref: &CommitRef) -> bool {
    job.display_name == af83::job_name(&commit_ref)
}

pub fn result_from_job(status: Option<String>) -> JobStatus {
    match status {
        None => JobStatus::Pending,
        Some(s) => {
            match s.as_ref() {
                "SUCCESS" => JobStatus::Success,
                "FAILURE" => JobStatus::Failure,
                _ => JobStatus::Unknown,
            }
        }
    }
}

pub fn jenkins_console_url_path(job_url: &String) -> String {
    format!("{}console", job_url)
}


fn jenkins_request_client(user_id: &String, token: &String) -> Result<reqwest::Client, Box<Error>> {
    let credentials = auth_credentials(user_id.to_owned(), token.to_owned());

    let mut headers = header::Headers::new();
    headers.set(header::Authorization(credentials));

    let client = reqwest::Client::builder()
        .default_headers(headers)
        .build()?;

    Ok(client)
}


#[cfg(test)]
mod tests {
    use self::mockito::mock;

    use super::*;

    fn test_request_client() -> reqwest::Client {
        jenkins_request_client(
            &"username".to_owned(),
            &"token".to_owned()
        ).expect("Failed to build Jenkins request client")
    }

    #[test]
    fn job_new_creates_a_job_from_payload() {
        let payload = r#"{
            "displayName": "3296-fix-typo-700d0",
            "result": "SUCCESS"
        }"#.to_owned();

        let job = Job::new(payload).expect("Failed to create job from payload");

        assert_eq!(job.display_name, "3296-fix-typo-700d0");
        assert_eq!(job.result, JobStatus::Success);
    }

    #[test]
    fn get_jobs_queries_jobs_from_jenkins_api() {
        let _mock = mock("GET", "/job/changes-branches/api/json")
            .with_status(200)
            .with_header("content-type", "application/json;charset=utf-8")
            .with_body(r#"
                {
                  "displayName": "changes-branches",
                  "builds": [
                    {
                      "_class": "hudson.model.FreeStyleBuild",
                      "number": 18,
                      "url": "http://jenkins.example.com/job/changes-branches/18/"
                    },
                    {
                      "_class": "hudson.model.FreeStyleBuild",
                      "number": 17,
                      "url": "http://jenkins.example.com/job/changes-branches/17/"
                    }
                  ]
                }
            "#)
            .create();

        let jobs = get_jobs(
            &mockito::SERVER_URL.to_owned(),
            &test_request_client(),
            "changes"
        ).expect("Failed to request jobs");

        assert_eq!(
            jobs,
            [
                "http://jenkins.example.com/job/changes-branches/18/",
                "http://jenkins.example.com/job/changes-branches/17/"
            ]
        );
    }

    #[test]
    fn request_job_queries_a_job_from_the_jenkins_api() {
        let _mock = mock("GET", "/job/changes-branches/15/api/json")
            .with_status(200)
            .with_header("content-type", "application/json;charset=utf-8")
            .with_body(r#"
                {
                  "displayName": "2388-delete-the-codes-391af",
                  "result": "SUCCESS"
                }
            "#)
            .create();

        let job = request_job(
            &mockito::SERVER_URL.to_owned(),
            &test_request_client(),
            "http://jenkins.example.com/job/changes-branches/15"
        ).expect("Failed to request job");

        let expected = Job {
            display_name: "2388-delete-the-codes-391af".to_owned(),
            result: JobStatus::Success,
        };

        assert_eq!(job.display_name, expected.display_name);
        assert_eq!(job.result, expected.result);
    }

    #[test]
    fn job_for_commit_returns_true_when_commit_matches_job() {
        let job = Job {
            display_name: "1753-fix-everything-b4a28".to_owned(),
            result: JobStatus::Pending,
        };

        let commit_ref = CommitRef {
            owner: "uso".to_owned(),
            repo: "vivid-system".to_owned(),
            sha: "b4a286e67d8f5eaf1f6ba5c57fc3c7d91ac0fd1c".to_owned(),
            branch: "1753-fix-everything".to_owned(),
        };

        assert_eq!(job_for_commit(&job, &commit_ref), true);
    }

    #[test]
    fn job_for_commit_returns_false_when_commit_doesnt_match_job() {
        let job = Job {
            display_name: "5234-eliminate-widgetmacallit-5a28c".to_owned(),
            result: JobStatus::Success,
        };

        let commit_ref = CommitRef {
            owner: "uso".to_owned(),
            repo: "vivid-system".to_owned(),
            sha: "b4a286e67d8f5eaf1f6ba5c57fc3c7d91ac0fd1c".to_owned(),
            branch: "1753-fix-everything".to_owned(),
        };

        assert_eq!(job_for_commit(&job, &commit_ref), false);
    }

    #[test]
    fn result_from_job_is_success() {
        assert_eq!(
            result_from_job(Some("SUCCESS".to_owned())),
            JobStatus::Success
        );
    }

    #[test]
    fn result_from_job_is_failure() {
        assert_eq!(
            result_from_job(Some("FAILURE".to_owned())),
            JobStatus::Failure
        );
    }

    #[test]
    fn result_from_job_is_pending() {
        assert_eq!(
            result_from_job(None),
            JobStatus::Pending
        );
    }

    #[test]
    fn jenkins_console_url_path_returns_url_to_console_page() {
        assert_eq!(
            jenkins_console_url_path(
                &"https://jenkins.example.com/job/changes-branches/15/".to_owned()
            ),
            "https://jenkins.example.com/job/changes-branches/15/console"
        );
    }
}


src/main.rs
// Copyright © 2017 Teddy Wing
//
// This file is part of Kipper.
//
// Kipper is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Kipper is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Kipper. If not, see <http://www.gnu.org/licenses/>.

extern crate getopts;
extern crate json;
#[macro_use]
extern crate log;
extern crate stderrlog;
#[macro_use]
extern crate rouille;

extern crate kipper;

use std::env;
use std::thread;
use std::io::Read;
use std::time::Duration;

use getopts::Options;

use kipper::jenkins;
use kipper::pull_request::{CommitRef, pull_request_opened_or_synchronized};

const DEFAULT_PORT: u16 = 8000;

fn print_usage(opts: Options) {
    let brief = "Usage: kipper --jenkins-url 'https://jenkins.example.com' --jenkins-user-id username --jenkins-token a72a57d448694703b2c3fd19e666ecc5 --github-token 1dc41fad0516460b870014b25b11847d";
    print!("{}", opts.usage(&brief));
}

fn internal_server_error() -> rouille::Response {
    rouille::Response::text("500 Internal Server Error")
        .with_status_code(500)
}

fn main() {
    let args: Vec<String> = env::args().collect();

    let mut opts = Options::new();
    opts.optopt("", "jenkins-url", "Jenkins URL (required)", "https://jenkins.example.com");
    opts.optopt("", "jenkins-user-id", "Jenkins user ID (required)", "USER_ID");
    opts.optopt("", "jenkins-token", "Jenkins API token (required)", "TOKEN");
    opts.optopt(
        "",
        "github-token",
        "GitHub API token with \"repo:status\" permission (required)",
        "TOKEN"
    );
    opts.optopt("p", "port", "set port number", "PORT");
    opts.optflag("h", "help", "print this help menu");

    let opt_matches = match opts.parse(&args[1..]) {
        Ok(m) => m,
        Err(e) => panic!(e.to_string()),
    };

    if opt_matches.opt_present("h") {
        print_usage(opts);
        return;
    }

    let jenkins_url = match opt_matches.opt_str("jenkins-url") {
        Some(url) => url,
        None => {
            print_usage(opts);
            return;
        },
    };

    let jenkins_user_id = match opt_matches.opt_str("jenkins-user-id") {
        Some(user_id) => user_id,
        None => {
            print_usage(opts);
            return;
        },
    };

    let jenkins_token = match opt_matches.opt_str("jenkins-token") {
        Some(token) => token,
        None => {
            print_usage(opts);
            return;
        },
    };

    let github_token = match opt_matches.opt_str("github-token") {
        Some(token) => token,
        None => {
            print_usage(opts);
            return;
        },
    };

    let port = match opt_matches.opt_str("p") {
        Some(p) => p.parse().expect("Unable to parse specified port"),
        None => DEFAULT_PORT,
    };

    // Logging
    stderrlog::new()
        .module(module_path!())
        .timestamp(stderrlog::Timestamp::Second)
        .verbosity(2) // LogLevel::Info
        .init()
        .expect("Logger failed to initialise");

    println!("Server listening on 0.0.0.0:{}", port);

    rouille::start_server(format!("0.0.0.0:{}", port), move |request| {
        router!(request,
            (POST) (/github/pull_request_event) => {
                match request.data() {
                    None => rouille::Response::text("400 Bad Request")
                        .with_status_code(400),
                    Some(mut data) => {
                        let mut body = String::new();
                        try_or_400!(data.read_to_string(&mut body));

                        let json = match json::parse(body.as_ref()) {
                            Ok(j) => j,
                            Err(e) => {
                                error!("{}", e.to_string());

                                return internal_server_error()
                            },
                        };

                        if !pull_request_opened_or_synchronized(json.clone()) {
                            return rouille::Response::text("No status update needed.")
                                .with_status_code(200)
                        }

                        let commit_ref = match CommitRef::new(json) {
                            Ok(cr) => cr,
                            Err(e) => {
                                error!("{}", e.to_string());

                                return internal_server_error()
                            },
                        };

                        // Clone so we can use these values in the thread
                        // closure. Since both closures are required to be
                        // 'static, we can't use references to these values.
                        let jenkins_url = jenkins_url.clone();
                        let jenkins_user_id = jenkins_user_id.clone();
                        let jenkins_token = jenkins_token.clone();
                        let github_token = github_token.clone();

                        thread::spawn(move || {
                            thread::sleep(Duration::from_secs(30));

                            match jenkins::find_and_track_build_and_update_status(
                                commit_ref,
                                jenkins_url,
                                &jenkins_user_id,
                                &jenkins_token,
                                github_token,
                            ) {
                                Ok(_) => {},
                                Err(e) => error!("{}", e.to_string()),
                            };
                        });

                        rouille::Response::text("202 Accepted")
                            .with_status_code(202)
                    }
                }
            },

            _ => rouille::Response::text("404 Not Found")
                    .with_status_code(404)
        )
    });
}
